File: src/lae/daemon/workspace_nav.py
"""Taskspace-scoped Hyprland workspace navigation."""

from __future__ import annotations

from lae.core.models import ContextMode, SessionState, Task
from lae.core.workspaces import (
    allowed_workspace_names,
    default_taskspace_workspace_names,
    task_workspace_names,
)
from lae.integrations import hyprland


def allowed_workspaces(state: SessionState) -> list[str]:
    return allowed_workspace_names(
        state, workspace_count=state.default_workspace_count
    )


def _relative_to_name(state: SessionState, relative: int) -> str | None:
    names = allowed_workspaces(state)
    if not names:
        return None
    idx = relative - 1
    if idx < 0 or idx >= len(names):
        return None
    return names[idx]


def _remember_workspace(state: SessionState, relative: int) -> None:
    state.last_workspace[state.taskspace_key()] = relative


def _active_relative(state: SessionState) -> int | None:
    active = hyprland.get_active_workspace()
    if not active or not active.name:
        return None
    names = allowed_workspaces(state)
    if active.name in names:
        return names.index(active.name) + 1
    return None


def workspace_go(state: SessionState, relative: int) -> str | None:
    name = _relative_to_name(state, relative)
    if name is None:
        return None
    hyprland.switch_workspace(name)
    _remember_workspace(state, relative)
    return name


def workspace_next(state: SessionState) -> str | None:
    names = allowed_workspaces(state)
    if not names:
        return None
    current_rel = _active_relative(state)
    if current_rel is not None:
        next_rel = (current_rel % len(names)) + 1
    else:
        next_rel = 1
    return workspace_go(state, next_rel)


def workspace_prev(state: SessionState) -> str | None:
    names = allowed_workspaces(state)
    if not names:
        return None
    current_rel = _active_relative(state)
    if current_rel is not None:
        prev_rel = current_rel - 1 if current_rel > 1 else len(names)
    else:
        prev_rel = len(names)
    return workspace_go(state, prev_rel)


def workspace_goto_name(state: SessionState, name: str) -> str | None:
    allowed = allowed_workspaces(state)
    if name not in allowed:
        return None
    hyprland.switch_workspace(name)
    if name in allowed:
        _remember_workspace(state, allowed.index(name) + 1)
    return name


def focus_last_workspace(state: SessionState) -> str | None:
    key = state.taskspace_key()
    relative = state.last_workspace.get(key, 1)
    name = _relative_to_name(state, relative)
    if name is None:
        names = allowed_workspaces(state)
        if not names:
            return None
        name = names[0]
        relative = 1
    hyprland.switch_workspace(name)
    _remember_workspace(state, relative)
    return name


def set_taskspace(
    state: SessionState,
    mode: ContextMode,
    task_id: str | None = None,
) -> None:
    if mode == ContextMode.task:
        if not task_id:
            raise ValueError("task_id required for task taskspace")
        if task_id not in state.tasks:
            raise ValueError(f"Unknown task: {task_id}")
        state.context_mode = ContextMode.task
        state.current_task_id = task_id
    elif mode == ContextMode.default:
        state.context_mode = ContextMode.default
        state.current_task_id = None
    focus_last_workspace(state)


def setup_task_workspaces(task: Task, *, slot_count: int | None = None) -> None:
    count = slot_count if slot_count is not None else task.workspace_count
    hyprland.ensure_workspaces(task_workspace_names(task.id, count))


def setup_default_taskspace_workspaces(count: int) -> None:
    hyprland.ensure_workspaces(default_taskspace_workspace_names(count))


# Deprecated aliases
set_context = set_taskspace
focus_last_desktop = focus_last_workspace
desktop_go = workspace_go
desktop_next = workspace_next
desktop_prev = workspace_prev
desktop_goto_name = workspace_goto_name
setup_default_workspaces = setup_default_taskspace_workspaces

File: crates/lae-core/src/workspace_nav.rs
//! Taskspace-scoped Hyprland workspace navigation.

use crate::hyprland;
use crate::models::{ContextMode, SessionState};
use crate::workspaces::{
    allowed_workspace_names, default_taskspace_workspace_names, task_workspace_names,
};

pub fn workspace_name_for_relative(state: &SessionState, relative: i32) -> Option<String> {
    relative_to_name(state, relative)
}

pub fn remember_workspace(state: &mut SessionState, relative: i32) {
    state
        .last_workspace
        .insert(state.taskspace_key(), relative);
}

pub fn workspace_next_relative(state: &SessionState) -> Option<i32> {
    let names = allowed_workspace_names(state);
    if names.is_empty() {
        return None;
    }
    match active_relative(state) {
        Some(current) => Some((current % names.len() as i32) + 1),
        None => Some(1),
    }
}

pub fn workspace_prev_relative(state: &SessionState) -> Option<i32> {
    let names = allowed_workspace_names(state);
    if names.is_empty() {
        return None;
    }
    match active_relative(state) {
        Some(current) if current > 1 => Some(current - 1),
        Some(_) => Some(names.len() as i32),
        None => Some(names.len() as i32),
    }
}

pub fn workspace_go(state: &mut SessionState, relative: i32) -> Option<String> {
    let name = workspace_name_for_relative(state, relative)?;
    hyprland::switch_workspace(&name);
    remember_workspace(state, relative);
    Some(name)
}

pub fn workspace_next(state: &mut SessionState) -> Option<String> {
    let relative = workspace_next_relative(state)?;
    workspace_go(state, relative)
}

pub fn workspace_prev(state: &mut SessionState) -> Option<String> {
    let relative = workspace_prev_relative(state)?;
    workspace_go(state, relative)
}

pub fn workspace_goto_name(state: &mut SessionState, name: &str) -> Option<String> {
    if state.context_mode != ContextMode::Global {
        let allowed = allowed_workspace_names(state);
        if !allowed.iter().any(|n| n == name) {
            return None;
        }
    }
    hyprland::switch_workspace(name);
    let allowed = allowed_workspace_names(state);
    if let Some(idx) = allowed.iter().position(|n| n == name) {
        remember_workspace(state, (idx + 1) as i32);
    }
    Some(name.to_string())
}

pub fn focus_last_workspace(state: &mut SessionState) -> Option<String> {
    let key = state.taskspace_key();
    let relative = *state.last_workspace.get(&key).unwrap_or(&1);
    if let Some(name) = relative_to_name(state, relative) {
        hyprland::switch_workspace(&name);
        remember_workspace(state, relative);
        return Some(name);
    }
    let names = allowed_workspace_names(state);
    let name = names.first()?.clone();
    hyprland::switch_workspace(&name);
    remember_workspace(state, 1);
    Some(name)
}

pub fn set_taskspace(state: &mut SessionState, mode: ContextMode, task_id: Option<&str>) -> Result<(), String> {
    match mode {
        ContextMode::Task => {
            let Some(task_id) = task_id else {
                return Err("task_id required for task taskspace".into());
            };
            if !state.tasks.contains_key(task_id) {
                return Err(format!("Unknown task: {task_id}"));
            }
            state.context_mode = ContextMode::Task;
            state.current_task_id = Some(task_id.to_string());
            state.previous_context = None;
            state.previous_task_id = None;
        }
        ContextMode::Default => {
            state.context_mode = ContextMode::Default;
            state.current_task_id = None;
            state.previous_context = None;
            state.previous_task_id = None;
        }
        ContextMode::Global => {
            state.previous_context = Some(state.context_mode);
            state.previous_task_id = state.current_task_id.clone();
            state.context_mode = ContextMode::Global;
        }
    }
    focus_last_workspace(state);
    Ok(())
}

pub fn toggle_global(state: &mut SessionState) {
    if state.context_mode == ContextMode::Global {
        restore_taskspace(state);
    } else {
        let _ = set_taskspace(state, ContextMode::Global, None);
    }
}

pub fn restore_taskspace(state: &mut SessionState) {
    let prev_mode = state.previous_context.unwrap_or(ContextMode::Default);
    let prev_task = state.previous_task_id.clone();
    state.previous_context = None;
    state.previous_task_id = None;
    if prev_mode == ContextMode::Task {
        if let Some(task_id) = prev_task {
            state.context_mode = ContextMode::Task;
            state.current_task_id = Some(task_id);
        } else {
            state.context_mode = ContextMode::Default;
            state.current_task_id = None;
        }
    } else {
        state.context_mode = ContextMode::Default;
        state.current_task_id = None;
    }
    focus_last_workspace(state);
}

pub fn setup_task_workspaces(task_id: &str, slot_count: u32) {
    hyprland::ensure_workspaces(&task_workspace_names(task_id, slot_count));
}

pub fn setup_task_workspaces_for_state(task_id: &str, state: &SessionState) {
    setup_task_workspaces(task_id, state.default_workspace_count);
}

pub fn setup_default_taskspace_workspaces(count: u32) {
    hyprland::ensure_workspaces(&default_taskspace_workspace_names(count));
}

fn relative_to_name(state: &SessionState, relative: i32) -> Option<String> {
    let names = allowed_workspace_names(state);
    let idx = (relative - 1) as usize;
    names.get(idx).cloned()
}

fn active_relative(state: &SessionState) -> Option<i32> {
    let active = hyprland::get_active_workspace().ok().flatten()?;
    if active.name.is_empty() {
        return None;
    }
    allowed_workspace_names(state)
        .iter()
        .position(|n| n == &active.name)
        .map(|i| (i + 1) as i32)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::models::SessionState;

    #[test]
    fn set_taskspace_default_clears_global_overlay() {
        let mut state = SessionState {
            context_mode: ContextMode::Global,
            previous_context: Some(ContextMode::Task),
            previous_task_id: Some("test-task".into()),
            current_task_id: Some("test-task".into()),
            default_workspace_count: 10,
            ..Default::default()
        };
        set_taskspace(&mut state, ContextMode::Default, None).unwrap();
        assert_eq!(state.context_mode, ContextMode::Default);
        assert!(state.previous_context.is_none());
        assert!(state.previous_task_id.is_none());
    }
}
